File: examples/python/bitwise-ops/BitwiseOps.runar.py
"""BitwiseOps -- Demonstrates bitwise and shift operators on Bigint values."""

from runar import SmartContract, Bigint, public, assert_


class BitwiseOps(SmartContract):
    """Demonstrates bitwise and shift operators."""

    a: Bigint
    b: Bigint

    def __init__(self, a: Bigint, b: Bigint):
        super().__init__(a, b)
        self.a = a
        self.b = b

    @public
    def test_shift(self):
        """Verify shift operators compile and run."""
        left = self.a << 2
        right = self.a >> 1
        assert_(left >= 0 or left < 0)
        assert_(right >= 0 or right < 0)
        assert_(True)

    @public
    def test_bitwise(self):
        """Verify bitwise operators compile and run."""
        and_result = self.a & self.b
        or_result = self.a | self.b
        xor_result = self.a ^ self.b
        not_result = ~self.a
        assert_(and_result >= 0 or and_result < 0)
        assert_(or_result >= 0 or or_result < 0)
        assert_(xor_result >= 0 or xor_result < 0)
        assert_(not_result >= 0 or not_result < 0)
        assert_(True)

File: examples/ruby/bitwise-ops/BitwiseOps.runar.rb
require 'runar'

# BitwiseOps -- Demonstrates bitwise and shift operators on Bigint values.

class BitwiseOps < Runar::SmartContract
  prop :a, Bigint, readonly: true
  prop :b, Bigint, readonly: true

  def initialize(a, b)
    super(a, b)
    @a = a
    @b = b
  end

  # Verify shift operators compile and run.
  runar_public
  def test_shift
    left = @a << 2
    right = @a >> 1
    assert left >= 0 || left < 0
    assert right >= 0 || right < 0
    assert true
  end

  # Verify bitwise operators compile and run.
  runar_public
  def test_bitwise
    and_result = @a & @b
    or_result = @a | @b
    xor_result = @a ^ @b
    not_result = ~@a
    assert and_result >= 0 || and_result < 0
    assert or_result >= 0 || or_result < 0
    assert xor_result >= 0 || xor_result < 0
    assert not_result >= 0 || not_result < 0
    assert true
  end
end
